File: Economy/client/economy.ts
import {
	BufReader,
	Items,
	ItemsMany,
	ItemsOne,
	type Quantity,
	SerialiseItem,
} from "./items"
import {
	Currency,
	IsMintable,
	Item,
	LimitedSource,
	Owner,
	UnlimitedSource,
	User,
} from "./types"

export const RandNumericId = (): number => Math.floor(Math.random() * 9e8) + 1e8

const idchars = "0123456789abcdefghijklmnopqrstuvwxyz"

export const RandStringId = (): string => {
	let result = ""
	for (let i = 0; i < 20; i++)
		result += idchars[Math.floor(Math.random() * idchars.length)]
	return result
}

export const ErrInvalidTransferID = new Error("invalid TransferID")

export class TransferID {
	constructor(
		private timestamp: bigint,
		private id: string
	) {}

	String(): string {
		return `${this.timestamp}-${this.id}`
	}

	Serialise(): Buffer {
		const bs = Buffer.alloc(8 + Buffer.byteLength(this.id))
		bs.writeBigUInt64BE(this.timestamp, 0)
		bs.write(this.id, 8)
		return bs
	}

	static Deserialise(data: Buffer): TransferID {
		if (data.length < 8) throw ErrInvalidTransferID
		const ts = data.readBigUInt64BE(0)
		const id = data.subarray(8).toString()
		return new TransferID(ts, id)
	}
}
// Send and Transfer types ported from ledger/economy.go
export class Send {
	constructor(
		public Owner: Owner | null,
		public Items: Items
	) {}

	String(): string {
		if (this.Owner === null) return `[] -> ${this.Items}`
		return `${this.Owner.String()} -> ${this.Items}`
	}

	Valid(): Error | null {
		for (const i of this.Items.One.set)
			if (this.Owner === null && !IsMintable(i))
				return new Error(`CanOwnOne ${i.String()} cannot be minted`)

		for (const [i, qty] of this.Items.Many.map) {
			if (qty === BigInt(0))
				return new Error(`CanOwnMany ${i.String()} has zero quantity`)
			if (this.Owner === null && !IsMintable(i))
				return new Error(`CanOwnMany ${i.String()} cannot be minted`)
		}
		return null
	}

	Equal(other: Send): boolean {
		if (this.Owner !== other.Owner) return false
		return this.Items.Equal(other.Items)
	}

	UnlimitedSourceAssetMint(): boolean {
		if (!(this.Owner instanceof UnlimitedSource)) return false
		const ous = this.Owner
		const expected = new Items(
			new ItemsOne([ous.Create()]),
			new ItemsMany()
		)
		return this.Items.Equal(expected)
	}

	Serialise(): Buffer {
		return Buffer.concat([
			// encode Owner
			SerialiseItem(this.Owner),
			// encode Items
			this.Items.Serialise(),
		])
	}

	static Deserialise(reader: BufReader): Send {
		const oi = Item.Deserialise(reader)
		let owner: Owner | null = null
		if (oi !== null)
			// runtime check
			owner = oi

		const items = Items.Deserialise(reader)
		return new Send(owner, items)
	}
}

export class Transfer {
	constructor(
		public Send0: Send,
		public Send1: Send
	) {}

	Serialise(): Buffer {
		return Buffer.concat([this.Send0.Serialise(), this.Send1.Serialise()])
	}

	static Deserialise(r: BufReader): Transfer {
		return new Transfer(Send.Deserialise(r), Send.Deserialise(r))
	}
}

export const ErrInvalidTransferWithID = new Error("invalid TransferWithID")

export class TransferWithID {
	constructor(
		public ID: TransferID,
		public Transfer: Transfer
	) {}

	String(): string {
		return `${this.ID.String()}: ${this.Transfer.Send0.String()}, ${this.Transfer.Send1.String()}`
	}

	static Deserialise(r: BufReader): TransferWithID {
		const l = r.readUint8()
		const idbuf = r.read(l)
		const tid = TransferID.Deserialise(idbuf)

		return new TransferWithID(tid, Transfer.Deserialise(r))
	}
}

export const ErrNoSourceOrDestination = new Error(
	"transfer has no source or destination"
)
export const ErrNoItems = new Error("transfer has no items")

export function TransferValid(t: Transfer): Error | null {
	if (t.Send0.Owner === null && t.Send1.Owner === null)
		return ErrNoSourceOrDestination
	if (t.Send0.Items.IsEmpty() && t.Send1.Items.IsEmpty()) return ErrNoItems

	const err0 = t.Send0.Valid()
	if (err0) return new Error(`invalid Send0: ${err0.message}`)
	const err1 = t.Send1.Valid()
	if (err1) return new Error(`invalid Send1: ${err1.message}`)

	return null
}

export function TransferSwap(t: Transfer): Transfer {
	return new Transfer(t.Send1, t.Send0)
}

function isSale(t: Transfer): boolean {
	const from = t.Send0.Owner
	if (!(from instanceof LimitedSource) && !(from instanceof UnlimitedSource))
		return false
	return t.Send1.Owner instanceof User
}

export function TransferSale(t: Transfer): boolean {
	if (t.Send0.Items.IsEmpty() && t.Send1.Items.IsEmpty()) return false
	return isSale(t) || isSale(TransferSwap(t))
}

function isStipend(t: Transfer): boolean {
	// A stipend has a nil owner and only currency in one direction, and a user on the other
	if (
		t.Send0.Owner !== null ||
		t.Send1.Owner === null ||
		t.Send0.Items.IsEmpty() ||
		!t.Send1.Items.IsEmpty()
	)
		return false
	if (t.Send0.Items.One.set.size > 0) return false
	if (t.Send0.Items.Many.map.size !== 1) return false
	if (!(t.Send1.Owner instanceof User)) return false
	for (const i of t.Send0.Items.Many.map.keys()) {
		return i instanceof Currency
	}
	return false
}

export function TransferStipend(t: Transfer): boolean {
	return isStipend(t) || isStipend(TransferSwap(t))
}

export function TransferEqual(a: Transfer, b: Transfer): boolean {
	return a.Send0.Equal(b.Send0) && a.Send1.Equal(b.Send1)
}

export function TransferSerialise(t: Transfer): Buffer {
	const b = []
	b.push(t.Send0.Serialise())
	b.push(t.Send1.Serialise())
	return Buffer.concat(b)
}

export function DeserialiseTransfer(buf: Buffer): Transfer {
	const r = new BufReader(buf)
	const s0 = Send.Deserialise(r)
	const s1 = Send.Deserialise(r)
	return new Transfer(s0, s1)
}

export class OwnersOne {
	set = new Set<Owner>()

	Add(o: Owner) {
		this.set.add(o)
	}

	Serialise(): Buffer {
		const b = []
		const lbuf = Buffer.alloc(4)
		lbuf.writeUInt32BE(this.set.size, 0)
		b.push(lbuf)
		for (const o of this.set) {
			b.push(SerialiseItem(o))
		}
		return Buffer.concat(b)
	}

	static Deserialise(r: BufReader): OwnersOne {
		const l = r.readUint32()
		const oo = new OwnersOne()
		for (let idx = 0; idx < l; idx++) {
			const i = Item.Deserialise(r)
			if (!(i instanceof Owner))
				throw new Error(`item is not Owner: ${i}`)
			oo.set.add(i)
		}
		return oo
	}
}

export class OwnersMany {
	map: Map<Owner, Quantity>

	constructor(initial?: Iterable<[Owner, Quantity]>) {
		this.map = new Map(initial)
	}

	Add(o: Owner, qty: Quantity) {
		const prev = this.map.get(o) ?? BigInt(0)
		this.map.set(o, prev + qty)
	}

	Serialise(): Buffer {
		const b = []
		const lbuf = Buffer.alloc(4)
		lbuf.writeUInt32BE(this.map.size, 0)
		b.push(lbuf)
		for (const [o, qty] of this.map) {
			b.push(SerialiseItem(o))
			const qtybuf = Buffer.alloc(8)
			qtybuf.writeBigUInt64BE(qty, 0)
			b.push(qtybuf)
		}
		return Buffer.concat(b)
	}

	static Deserialise(reader: BufReader): OwnersMany {
		const l = reader.readUint32()
		const om = new OwnersMany()
		for (let idx = 0; idx < l; idx++) {
			const i = Item.Deserialise(reader)
			if (!(i instanceof Owner))
				throw new Error(`item is not Owner: ${i}`)
			const qty = reader.readUint64()
			om.map.set(i, qty)
		}
		return om
	}
}

File: Economy/ledger/economy.go
package ledger

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math/rand"
	"time"

	gonanoid "github.com/matoous/go-nanoid/v2"
)

func RandNumericId() (id uint32) {
	return uint32(rand.Intn(9e8) + 1e8) // 9 digit
}

const idchars = "0123456789abcdefghijklmnopqrstuvwxyz"

func RandStringId() (id string) {
	id, _ = gonanoid.Generate(idchars, 20) // doesn't error at runtime, really
	return
}

// assets are the worst grey area ever
// they are fungible, because users can own (as in have bought) multiple of the same asset ID
// but also not, because only one user can own (as in being the seller) of an asset
// HERE'S MY SOLUTION
// we have 2 item types associated with assets: the asset itself, and the source of the asset
// with the asset itself being fungible and the source being non-fungible

type TransferID struct {
	timestamp uint64
	id        string
}

func (t TransferID) String() string {
	return fmt.Sprintf("%d-%s", t.timestamp, t.id)
}

func MakeTransferID() TransferID {
	return TransferID{
		timestamp: uint64(time.Now().UnixNano()),
		id:        RandStringId(),
	}
}

// The timestamp comes first so they are ordered chronologically when sorted lexicographically
func (t TransferID) Serialise() []byte {
	bs := make([]byte, 8+len(t.id))
	binary.BigEndian.PutUint64(bs[:8], t.timestamp)
	copy(bs[8:], t.id)
	return bs
}

var ErrInvalidTransferID = errors.New("invalid TransferID")

func DeserialiseTransferID(data []byte) (t TransferID, err error) {
	if len(data) < 8 {
		return TransferID{}, ErrInvalidTransferID
	}
	t.timestamp = binary.BigEndian.Uint64(data[:8])
	t.id = string(data[8:])
	return
}

// A Send represents items being sent FROM a user/group
type Send struct {
	Owner Owner // again no embed pls (not as much of an issue here)
	Items Items
}

func (s Send) String() string {
	if s.Owner == nil {
		return fmt.Sprintf("[] -> %v", s.Items)
	}
	return fmt.Sprintf("%s -> %v", s.Owner, s.Items)
}

func (s Send) Valid() error {
	for i := range s.Items.One {
		if s.Owner == nil && !IsMintable(i) {
			return fmt.Errorf("CanOwnOne %v cannot be minted", i)
		}
	}
	for i, qty := range s.Items.Many {
		if qty == 0 {
			return fmt.Errorf("CanOwnMany %v has zero quantity", i)
		}
		if s.Owner == nil && !IsMintable(i) {
			return fmt.Errorf("CanOwnMany %v cannot be minted", i)
		}
	}
	return nil
}

func (s Send) Equal(other Send) bool {
	if s.Owner != other.Owner {
		return false
	}
	return s.Items.Equal(other.Items)
}

func (s Send) UnlimitedSourceAssetMint() bool {
	ous, ok := s.Owner.(UnlimitedSource)
	if !ok {
		return false
	}

	return s.Items.Equal(Items{
		One: ItemsOne{
			ous.Create(): {},
		},
	})
}

func (s Send) Serialise(w io.Writer) {
	// encode Owner
	if !SerialiseItem(s.Owner, w) {
		panic(fmt.Sprintf("unknown Owner type: %T", s.Owner))
	}

	// encode Items
	if err := s.Items.Serialise(w); err != nil {
		panic(fmt.Sprintf("serialise Items: %v", err))
	}
}

func DeserialiseSend(r io.Reader) (Send, error) {
	var s Send

	// decode Owner
	oi, err := DeserialiseItem(r)
	if err != nil {
		return s, fmt.Errorf("decode owner: %w", err)
	}

	if oi != nil {
		o, ok := oi.(Owner)
		if !ok {
			return s, fmt.Errorf("item is not Owner: %T", oi)
		}
		s.Owner = o
	}

	// decode Items
	s.Items, err = DeserialiseItems(r)
	if err != nil {
		return s, fmt.Errorf("decode items: %w", err)
	}

	return s, nil
}

// A transfer with a From and To is a transaction
// A transfer with only a From is a burn
// A transfer with only a To is a mint
// A transfer with neither is INVALID!
// The empty UserID refers to the nil account for mints/burns
type Transfer [2]Send

func (t Transfer) String() string {
	return fmt.Sprintf("%s, %s", t[0], t[1])
}

type TransferWithID struct {
	ID       TransferID
	Transfer Transfer
}

func (t TransferWithID) String() string {
	return fmt.Sprintf("%s: %s", t.ID, t.Transfer)
}

var (
	ErrNoSourceOrDestination = errors.New("transfer has no source or destination")
	ErrNoItems               = errors.New("transfer has no items")
)

func (t Transfer) Valid() error {
	// Can't have a transfer of nothing or between nobody
	if t[0].Owner == nil && t[1].Owner == nil {
		return ErrNoSourceOrDestination
	}
	if t[0].Items.IsEmpty() && t[1].Items.IsEmpty() {
		return ErrNoItems
	}
	for i, send := range t {
		if err := send.Valid(); err != nil {
			return fmt.Errorf("invalid Send[%d]: %w", i, err)
		}
	}
	return nil
}

func (t Transfer) Swap() Transfer {
	return Transfer{t[1], t[0]}
}

func isSale(t Transfer) bool {
	switch t[0].Owner.(type) {
	case LimitedSource, UnlimitedSource:
	default:
		return false
	}

	_, ok := t[1].Owner.(User)
	return ok
}

// This is how an angel dies
// I blame it on my own sick pride
// Blame it on my ADD, baby
func (t Transfer) Sale() bool {
	// items could be empty on one side if the item is free or smth
	if t[0].Items.IsEmpty() && t[1].Items.IsEmpty() {
		return false
	}

	return isSale(t) || isSale(t.Swap())
}

func isStipend(t Transfer) bool {
	// A stipend has a nil owner and only currency in one direction, and a user on the other
	if t[0].Owner != nil || t[1].Owner == nil || t[0].Items.IsEmpty() || !t[1].Items.IsEmpty() || len(t[0].Items.One) > 0 || len(t[0].Items.Many) != 1 {
		return false
	}

	if _, ok := t[1].Owner.(User); !ok {
		return false
	}

	for i := range t[0].Items.Many {
		_, ok := i.(Currency)
		return ok // there's only one item here
	}

	return false // never reached i guess
}

func (t Transfer) Stipend() bool {
	return isStipend(t) || isStipend(t.Swap())
}

func (t Transfer) Equal(other Transfer) bool {
	return t[0].Equal(other[0]) && t[1].Equal(other[1])
}

func (t Transfer) Serialise(w io.Writer) {
	t[0].Serialise(w)
	t[1].Serialise(w)
}

func DeserialiseTransfer(r io.Reader) (t Transfer, err error) {
	if t[0], err = DeserialiseSend(r); err != nil {
		return Transfer{}, fmt.Errorf("decode Send[0]: %w", err)
	}

	if t[1], err = DeserialiseSend(r); err != nil {
		return Transfer{}, fmt.Errorf("decode Send[1]: %w", err)
	}

	return
}

type OwnersOne map[Owner]struct{}

func (oo *OwnersOne) Add(o Owner) {
	if *oo == nil {
		*oo = OwnersOne{o: struct{}{}}
		return
	}
	(*oo)[o] = struct{}{}
}

func (oo OwnersOne) Serialise(w io.Writer) error {
	var lbuf [4]byte
	binary.BigEndian.PutUint32(lbuf[:], uint32(len(oo)))
	w.Write(lbuf[:])

	for o := range oo {
		if !SerialiseItem(o, w) {
			return fmt.Errorf("unknown Owner type: %T", o)
		}
	}
	return nil
}

func DeserialiseOwnersOne(r io.Reader) (OwnersOne, error) {
	var lbuf [4]byte
	if _, err := r.Read(lbuf[:]); err != nil {
		return nil, fmt.Errorf("read OwnersOne length: %w", err)
	}
	l := binary.BigEndian.Uint32(lbuf[:])

	oo := make(OwnersOne, l)
	for range l {
		i, err := DeserialiseItem(r)
		if err != nil {
			return nil, fmt.Errorf("decode owner: %w", err)
		}
		o, ok := i.(Owner)
		if !ok {
			return nil, fmt.Errorf("item is not Owner: %T", i)
		}
		oo[o] = struct{}{}
	}
	return oo, nil
}

type OwnersMany map[Owner]Quantity

func (om *OwnersMany) Add(o Owner, qty Quantity) {
	if *om == nil {
		*om = OwnersMany{o: qty}
		return
	}
	(*om)[o] += qty
}

func (om OwnersMany) Serialise(w io.Writer) error {
	var lbuf [4]byte
	binary.BigEndian.PutUint32(lbuf[:], uint32(len(om)))
	w.Write(lbuf[:])

	for o, qty := range om {
		if !SerialiseItem(o, w) {
			return fmt.Errorf("unknown Owner type: %T", o)
		}

		var qtybuf [8]byte
		binary.BigEndian.PutUint64(qtybuf[:], uint64(qty))
		w.Write(qtybuf[:])
	}
	return nil
}

func DeserialiseOwnersMany(r io.Reader) (OwnersMany, error) {
	var lbuf [4]byte
	if _, err := r.Read(lbuf[:]); err != nil {
		return nil, fmt.Errorf("read OwnersMany length: %w", err)
	}
	l := binary.BigEndian.Uint32(lbuf[:])

	om := make(OwnersMany, l)
	for range l {
		i, err := DeserialiseItem(r)
		if err != nil {
			return nil, fmt.Errorf("decode owner: %w", err)
		}
		o, ok := i.(Owner)
		if !ok {
			return nil, fmt.Errorf("item is not Owner: %T", i)
		}

		var qtybuf [8]byte
		if _, err := r.Read(qtybuf[:]); err != nil {
			return nil, fmt.Errorf("read quantity: %w", err)
		}
		qty := binary.BigEndian.Uint64(qtybuf[:])
		om[o] = Quantity(qty)
	}
	return om, nil
}
